From 9cbcdcaf92546dd289a5a6ef276bd6c34ed12d70 Mon Sep 17 00:00:00 2001
From: Mario Schuknecht <mario.schuknecht@dresearch-fe.de>
Date: Tue, 4 Sep 2018 11:12:46 +0200
Subject: [PATCH] pixcir touch: moving mode HYP-19464

To enable wipe gestures set interrupt mode in R52 to 0x1 'assert only
when finger moving'. The event triggering was adapted for this.

Signed-off-by: Mario Schuknecht <mario.schuknecht@dresearch-fe.de>
---
 drivers/input/touchscreen/pixcir_i2c_ts.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/drivers/input/touchscreen/pixcir_i2c_ts.c b/drivers/input/touchscreen/pixcir_i2c_ts.c
index 3bb0637d832e..a7a2a2a91e9c 100644
--- a/drivers/input/touchscreen/pixcir_i2c_ts.c
+++ b/drivers/input/touchscreen/pixcir_i2c_ts.c
@@ -159,7 +159,13 @@ static irqreturn_t pixcir_ts_isr(int irq, void *dev_id)
 		pixcir_ts_report(tsdata, &report);
 
 		if (gpiod_get_value_cansleep(tsdata->gpio_attb)) {
-			if (report.num_touches) {
+			if (!report.num_touches) {
+				/*
+				 * Finger up? Then synchronize now. This is required
+				 * in interrupt mode 0x1 PIXCIR_INT_DIFF_COORD
+				 * only when finger moving. MST
+				 */
+				input_mt_drop_unused(tsdata->input);
 				/*
 				 * Last report with no finger up?
 				 * Do it now then.
@@ -303,7 +309,8 @@ static int pixcir_start(struct pixcir_i2c_ts_data *ts)
 	}
 
 	/* LEVEL_TOUCH interrupt with active low polarity */
-	error = pixcir_set_int_mode(ts, PIXCIR_INT_LEVEL_TOUCH, 0);
+	/* Use INT_MODE 0x1 only when finger moving. MST */
+	error = pixcir_set_int_mode(ts, PIXCIR_INT_DIFF_COORD, 0);
 	if (error) {
 		dev_err(dev, "Failed to set interrupt mode: %d\n", error);
 		return error;
@@ -506,8 +513,10 @@ static int pixcir_i2c_ts_probe(struct i2c_client *client,
 			 tsdata->max_fingers);
 	}
 
+	/* For INT_MODE 0x1 PIXCIR_INT_DIFF_COORD INPUT_MT_DROP_UNUSED does not
+	 * have to be activated here. MST */
 	error = input_mt_init_slots(input, tsdata->max_fingers,
-				    INPUT_MT_DIRECT | INPUT_MT_DROP_UNUSED);
+				    INPUT_MT_DIRECT);
 	if (error) {
 		dev_err(dev, "Error initializing Multi-Touch slots\n");
 		return error;
-- 
2.16.4

